src/V3Clock.cpp

```diff
@@ -151,12 +175,58 @@ class ClockVisitor final : public VNVisitor {
         clearLastSen();
     }
 
+    // Create sampled values
+    virtual void visit(AstNodeModule* nodep) override {
+        VL_RESTORER(m_modp);
+        {
+            m_modp = nodep;
+            iterateChildren(nodep);
+        }
+    }
+    virtual void visit(AstTopScope* nodep) override {
+        VL_RESTORER(m_scopep);
+        {
+            m_scopep = nodep->scopep();
+            iterateChildren(nodep);
+        }
+    }
+    virtual void visit(AstScope* nodep) override {
+        VL_RESTORER(m_scopep);
+        {
+            m_scopep = nodep;
+            iterateChildren(nodep);
+        }
+    }
+    virtual void visit(AstSampled* nodep) override {
+        VL_RESTORER(m_inSampled);
+        {
+            m_inSampled = true;
+            iterateChildren(nodep);
+            nodep->replaceWith(nodep->exprp()->unlinkFrBack());
+            VL_DO_DANGLING(pushDeletep(nodep), nodep);
+        }
+    }
+    virtual void visit(AstVarRef* nodep) override {
+        iterateChildren(nodep);
+        if (m_inSampled && !nodep->user1SetOnce()) {
+            AstVarScope* const varscp = nodep->varScopep();
+            AstVarScope* const lastscp = createSampledVar(varscp);
+            AstNode* const newp = new AstVarRef(nodep->fileline(), lastscp, VAccess::READ);
```

Comment:
```suggestion
            AstNode* const newp = new AstVarRef{nodep->fileline(), lastscp, VAccess::READ};
```
Also, is it possible for `nodep` to be `WRITE`? If it doesn't error out earlier, add a `v3error` here. If it does, add an assert here.

Reply:
I errored it out, but it's an interesting case. The grammar of Verilator only allows expressions there, so no value can be modified except for incrementations and decrementations.

Since values accessed in asserts are sampled, it seems unclear what should happen when one of these two is used.

Reply:
Nice catch. So does the LRM allow incrementations/decrementations there? If not, we should make it clear it's illegal and add a test. If it does, and handling that case is difficult, we should warn that it's unsupported. This probably won't be caught at this stage, as various transformations along the way would make it into a normal expression.

What about output function arguments and non-pure functions? Seems like there's huge potential for bugs. Maybe we should also treat function calls in `$sampled` as unsupported.

Please add tests for these edge cases (unless they turn out to be trivial).

Reply:
Okay, so I made a few changes:
* writes to variables in sampled areas stay errored out as unsupported,
* pure functions work well now,
* function side effects have undefined behavior,
* sampling in concurrent asserts is injected on parsing stage.

I also made a few clean ups.

To sum up:
Our problems came from the fact that we want $sampled to be eventually applied to variables only, but optimizations can break it in two ways:
1. Moving a sampled variable from inside a sampled expression to another place (like it was with function arguments).
2. Wrapping a $sampled variable into something extra, thus breaking the desired state where only variables are sampled.

Now we're ready for 1. because at an early stage (V3Assert) we simplify sampled expressions to ones that only have sampled variables. So in case an optimization moves a part of an expression somewhere else, all variables go with their $sampled.

We're also ready for 2. because at a late stage (V3Clock) where calls to $sampled are replaced by helper variables, we don't care if the argument is actually a variable, we just recursively look for every variable inside the call, and replace it with a helper.

Reply:
Sounds good. See my other remaining comment.